presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseToMap.java

```diff
@@ -0,0 +1,278 @@
+            return RowExpressionTreeRewriter.rewriteWith(caseToMapRewriter, expression);
+        }
+    }
+
+    private static class CaseToMapRewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+        private final LogicalRowExpressions logicalRowExpressions;
+
+        private CaseToMapRewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+            this.logicalRowExpressions = new LogicalRowExpressions(
+                    new RowExpressionDeterminismEvaluator(functionAndTypeManager),
+                    functionResolution,
+                    functionAndTypeManager);
+        }
+
+        private boolean addKeyValue(RowExpression key, Set<RowExpression> keySet, List<RowExpression> keys, RowExpression value, List<RowExpression> values)
+        {
+            // matching types and non-null values only allowed
+            if (!(key instanceof ConstantExpression) ||
+                    ((ConstantExpression) key).getValue() == null ||
+                    (keys.size() > 0 && !keys.get(0).getType().equals(key.getType()))) {
+                return false;
+            }
+
+            if (keySet.add(key)) {
+                // We allow all same type only
+                if (values.size() > 0 && !values.get(0).getType().equals(value.getType())) {
+                    return false;
+                }
+
+                keys.add(key);
+                values.add(value);
+            }
+
+            return true;
+        }
+
+        @Override
+        public RowExpression rewriteSpecialForm(SpecialFormExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getForm() != SWITCH) {
+                return rewriteRowExpression(node, context, treeRewriter);
+            }
+
+            // by construction we should have at least one WHEN and ELSE is always added if misseing
+            int numArgs = node.getArguments().size();
+            RowExpression lastArg = node.getArguments().get(numArgs - 1);
+
+            checkState(numArgs >= 2);
+            checkState(!(lastArg instanceof SpecialFormExpression && ((SpecialFormExpression) lastArg).getForm().equals(WHEN)));
+
+            if (!(lastArg instanceof ConstantExpression)) {
+                return node;
+            }
+
+            RowExpression firstArg = node.getArguments().get(0);
+            Set<RowExpression> keySet = new HashSet<>();
+            List<RowExpression> whens = new ArrayList<RowExpression>(node.getArguments().size());
+            List<RowExpression> thens = new ArrayList<RowExpression>(node.getArguments().size());
+            RowExpression checkExpr;
+            int start;
```

Comment:
can we initialize checkExpr = null and start = 0 and compress some of the branches below:
```
if (!(firstArg instanceof SpecialFormExpression && ((SpecialFormExpression) firstArg).getForm().equals(WHEN))) {
    if (!firstArg.equals(constant(true, BOOLEAN))) {
        checkExpr = firstArg;
    }

    start = 1;
}
```

Reply:
It will be unnecessarily overwriting in many cases :)

Reply:
@kaikalur I think its better to simplify the code here, than try to do less overwrites, which may not make a difference

Reply:
Yeah -  I think this is fine. It's more readable